[scripts] fixes for armv7l host builds

vserdyuk opened the pull request:
Fixes some problems when using vcpkg on host armv7l/aarch64. It allows to build qtdeclarative shared libraries on the target machine (unfortunately, cross-compiling Qt on x86_x64 host for armv7l target does not work for me, but that's another issue).

The proper fix for Ninja is to respect VCPKG_FORCE_SYSTEM_BINARIES (see #21482 and comment https://github.com/microsoft/vcpkg/pull/23659#issuecomment-1151867976), but it will probably need some time to be implemented. So this fix will work until then.

- #### What does your PR fix?
  Fixes #25277

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  No

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  I guess yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  n/a

vserdyuk committed:
[scripts] fixes for armv7l host builds

vserdyuk committed:
[scripts] fix meson for 32-bit arm

Osyotr commented on scripts/cmake/vcpkg_find_acquire_program.cmake:
```diff
@@ -550,9 +555,13 @@ function(vcpkg_find_acquire_program program)
         set(program_version 0.14.5)
         set(supported_on_unix ON)
         if(CMAKE_HOST_SYSTEM_NAME STREQUAL "Linux")
+            EXEC_PROGRAM(uname ARGS -m OUTPUT_VARIABLE HOST_ARCH)
```
HOST_ARCH is already set

Osyotr replied:
Ah sorry, didn't read the comment

vserdyuk replied:
Yes, it's defined in each `program` branch separately (when needed). Should probably be defined once in the beginning of the function, though.

Osyotr commented on scripts/cmake/vcpkg_configure_meson.cmake:
```diff
@@ -224,6 +224,9 @@ function(z_vcpkg_meson_generate_cross_file additional_binaries) #https://mesonbu
             set(build_cpu armv8)
+        elseif(MACHINE MATCHES "armv7h?l")
+            set(build_cpu_fam arm)
+            set(build_cpu ${MACHINE})
```
This code should probably match code for host detection below

vserdyuk replied:
If you mean `VCPKG_TARGET_ARCHITECTURE` check, I think those variables are different. `VCPKG_TARGET_ARCHITECTURE` is defined by vcpkg and for ARM family it is either `arm` or `arm64` while `MACHINE ` variable is a `uname -m` call result and can be `aarch64`, `armv7l`, `armv7h`, etc.

Osyotr replied:
I mean, is it ok to assume that host CPU is always `armv7hl` if build CPU can be either `armv7l` or `armv7hl`

vserdyuk replied:
Unfortunately, the `arm` value of `VCPKG_TARGET_ARCHITECTURE` lost it's precision whether it's `armv7l` or `armv7hl` while `MACHINE` variable has the exact information. Now we do our best effort to provide the most accurate information.
I have 32-bit Ubuntu 20.04 on Raspberry Pi 4B and `uname -m` gives me `armv7l`. So `build_cpu` being `armv7l` works with `host_cpu` being `armv7hl`. Haven't checked whether it will work if I change to `set(build_cpu armv7hl)`. Do you think it's better to hard code it this way?

Cheney-W commented:
The failure of `kf5globalaccel` and `libx11` is fixing by PR [26486](https://github.com/microsoft/vcpkg/pull/26486).

Neumann-A commented:
`kf5globalaccel` is not fixed by that. Seems like `kf5globalaccel` needs to explicitly disable x11 usage on windows.

vserdyuk committed:
Merge branch 'microsoft:master' into master

vserdyuk commented:
Are there any additional steps to be taken to push this forward? Considering failing pipelines (Windows. UWP, etc.) have nothing to do with the change?

Cheney-W commented:
/azp run

Neumann-A commented:
Can you not simply use `set(NINJA <yourpath>)` in the triplet to achieve the same ?

dg0yt commented on scripts/cmake/vcpkg_find_acquire_program.cmake:
```diff
@@ -313,13 +314,17 @@ function(vcpkg_find_acquire_program program)
-            set(tool_subdirectory "${program_version}-linux")
-            set(paths_to_search "${DOWNLOADS}/tools/ninja-${program_version}-linux")
-            set(download_sha512 93e802e9c17fb59636cddde4bad1ddaadad624f4ecfee00d5c78790330a4e9d433180e795718cda27da57215ce643c3929cf72c85337ee019d868c56f2deeef3)
+            EXEC_PROGRAM(uname ARGS -m OUTPUT_VARIABLE HOST_ARCH)
```
`exec_program` is deprecated already since CMake 3.0.

dg0yt commented on scripts/cmake/vcpkg_find_acquire_program.cmake:
```diff
@@ -550,9 +555,13 @@ function(vcpkg_find_acquire_program program)
         set(program_version 0.14.5)
         set(supported_on_unix ON)
         if(CMAKE_HOST_SYSTEM_NAME STREQUAL "Linux")
+            EXEC_PROGRAM(uname ARGS -m OUTPUT_VARIABLE HOST_ARCH)
```
`exec_program` is deprecated already since CMake 3.0.

vserdyuk replied:
I replaced it with execute_process

vserdyuk commented:
> Can you not simply use `set(NINJA <yourpath>)` in the triplet to achieve the same ?

I reverted my fix and added hard-coded path to ninja this to arm-linux triplet and it worked. But if we take this approach, we will have to patch all non-x64 linux tripet files.

vserdyuk committed:
[scripts] replaced EXEC_PROGRAM with execute_process in ninja and pathelf paths in vcpkg_find_acquire_program

dg0yt commented:
> But if we take this approach, we will have to patch all non-x64 linux tripet files.

No, it is a host tool, so in general, its path cannot be in the (target) triplet files.

Cheney-W commented:
/azp run
